swarm/storage/hasherstore.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/ethereum/go-ethereum/swarm/storage/encryption"
+)
+
+type chunkEncryption struct {
+	spanEncryption encryption.Encryption
+	dataEncryption encryption.Encryption
+}
+
+type hasherStore struct {
+	store           ChunkStore
+	hashFunc        SwarmHasher
+	chunkEncryption *chunkEncryption
+	refSize         int64
+	wg              *sync.WaitGroup
+	closed          chan struct{}
+}
+
+func newChunkEncryption() *chunkEncryption {
+	return &chunkEncryption{
+		spanEncryption: encryption.New(0, math.MaxUint32, sha3.NewKeccak256),
+		dataEncryption: encryption.New(4096, 0, sha3.NewKeccak256),
+	}
+}
+
+// NewHasherStore creates a hasherStore object, which implements Putter and Getter interfaces.
+// With the HasherStore you can put and get chunk data (which is just []byte) into a ChunkStore
+// and the hasherStore will take core of encryption/decryption of data if necessary
+func NewHasherStore(chunkStore ChunkStore, hashFunc SwarmHasher, toEncrypt bool) *hasherStore {
+	var chunkEncryption *chunkEncryption
+
+	refSize := int64(hashFunc().Size())
+	if toEncrypt {
+		chunkEncryption = newChunkEncryption()
+		refSize += encryption.KeyLength
+	}
+
+	return &hasherStore{
+		store:           chunkStore,
+		hashFunc:        hashFunc,
+		chunkEncryption: chunkEncryption,
+		refSize:         refSize,
+		wg:              &sync.WaitGroup{},
+		closed:          make(chan struct{}),
+	}
+}
+
+// Puts the chunkData into the ChunkStore of the hasherStore and returns the reference.
+// If hasherStore has a chunkEncryption object, the data will be encrypted.
+// Asynchronous function, the data will not necessarily be stored when it returns.
+func (h *hasherStore) Put(chunkData ChunkData) (Reference, error) {
+	c := chunkData
+	size := chunkData.Size()
+	var encryptionKey encryption.Key
+	if h.chunkEncryption != nil {
+		var err error
+		c, encryptionKey, err = h.encryptChunkData(chunkData)
+		if err != nil {
+			return nil, err
+		}
+	}
+	chunk := h.createChunk(c, size)
+
+	h.storeChunk(chunk)
+
+	return Reference(append(chunk.Key, encryptionKey...)), nil
+}
+
+// Returns data of the chunk with the given reference (retrieved from the ChunkStore of hasherStore).
+// If the data is encrypted and the reference contains an encryption key, it will be decrypted before
+// return.
+func (h *hasherStore) Get(ref Reference) (ChunkData, error) {
+	if h.store == nil {
+		return nil, errors.New("Can not get ref from HasherStore with nil ChunkStore")
+	}
+	key, encryptionKey, err := parseReference(ref)
+	if err != nil {
+		return nil, err
+	}
+	toDecrypt := (encryptionKey != nil)
+
+	chunk, err := h.store.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	chunkData := chunk.SData
+	if toDecrypt {
+		var err error
+		chunkData, err = h.decryptChunkData(chunkData, encryptionKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return chunkData, nil
+}
+
+// The Close() indicates that no more chunks will be put with the hasherStore, so the Wait
```

Comment:
Close indicates...

Reply:
Fixed in 2329cfb50e285f49e3c4f59eb6f6e5ce09acac27